File: C++/suffix_array.cc
#include <iostream>
#include <cstring>
#include <vector>
#include <algorithm>

using namespace std;

class SuffixArray {
    private:
        struct suffix {
            int index; 
            int rank[2];
        };

        static int cmp(suffix &a, suffix &b) {
            return (a.rank[0] == b.rank[0])? (a.rank[1] < b.rank[1] ?1: 0):
                (a.rank[0] < b.rank[0] ?1: 0);
        }

        int n;
        vector<int> suffixArr;
        string origin;

    public:
        int buildSuffixArray(string txt) {
            origin = txt;
            n = txt.length();
            int ind[n];  
            vector<suffix> suffixes;

            for (int i = 0; i < n; i++) {
                suffixes.push_back({i, {txt[i], i < n - 1 ? txt[i + 1]: -1}});
            }
            sort(suffixes.begin(), suffixes.end(), cmp);
            for (int k = 2; k < n; k = k*2) {
                int rank = 0;
                int prev_rank = suffixes[0].rank[0];
                suffixes[0].rank[0] = rank;
                ind[suffixes[0].index] = 0;

                for (int i = 1; i < n; i++) {
                    if (suffixes[i].rank[0] == prev_rank && suffixes[i].rank[1] == suffixes[i-1].rank[1]) {
                        suffixes[i].rank[0] = rank;
                    } else { 
                        prev_rank = suffixes[i].rank[0];
                        suffixes[i].rank[0] = ++rank;
                    }
                    ind[suffixes[i].index] = i;
                }
                for (int i = 0; i < n; i++) {
                    int nextindex = suffixes[i].index + k;
                    suffixes[i].rank[1] = (nextindex < n)?
                        suffixes[ind[nextindex]].rank[0]: -1;
                }
                sort(suffixes.begin(), suffixes.end(), cmp);
            }
            suffixArr.clear();
           for (int i = 0; i < n; i++)
               suffixArr.push_back(suffixes[i].index);
        }

        int search(string substr) {
            int left = 0, right = n - 1;
            int m = substr.length();

            while (left <= right) {
                int mid = (left + right) / 2;
                int res = substr.compare(origin.substr(suffixArr[mid], m));
                if (res == 0) return suffixArr[mid];
                if (res < 0) right = mid - 1; else left = mid + 1;
            }
            return -1;
        }

        // Print SuffixArray
        void printArr() {
            for (int i = 0; i < suffixArr.size(); i++)
                cout << suffixArr[i] << " ";
            cout << endl;
        }
};

int main() {
    string txt = "banana";

    SuffixArray s;
    s.buildSuffixArray(txt);
    s.printArr();
    cout << s.search("ana") << endl;
    cout << s.search("nan") << endl;
    cout << s.search("ann") << endl;
    cout << s.search("bb") << endl;
    return 0;
}

File: Java/SuffixArray.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;


public class SuffixArray {

  private int[] indexArray;
  private String origin;

  static class Compare implements Comparator<int[]> {

    @Override
    public int compare(int[] x, int[] y) {
      return x[1] == y[1] ? x[2] - y[2] : x[1] - y[1];
    }
  }

  public void build(String txt) {
    origin = txt;
    int n = txt.length();
    List<int[]> suffix = new ArrayList<>();
    int[] pos = new int[n];

    for (int i = 0; i < n; i++) {
      int[] temp = {i, txt.charAt(i) - 'a', i < n - 1 ? txt.charAt(i + 1) : -1};
      suffix.add(temp);
    }
    suffix.sort(new Compare());
    for (int k = 2; k < n; k = k * 2) {
      int rank = 0;
      int prev_rank = -1;
      pos[suffix.get(0)[0]] = 0;
      for (int i = 1; i < n; i++) {
        if (suffix.get(i)[1] == prev_rank && suffix.get(i)[2] == suffix.get(i - 1)[2]) {
          suffix.get(i)[1] = rank;
        } else {
          prev_rank = suffix.get(i)[1];
          suffix.get(i)[1] = ++rank;
        }
        pos[suffix.get(i)[0]] = i;
      }
      for (int i = 0; i < n; i++) {
        int index = suffix.get(i)[0];
        suffix.get(i)[2] = index + k < n ? suffix.get(pos[index + k])[1] : -1;
      }
      suffix.sort(new Compare());
    }
    indexArray = new int[n];
    for (int i = 0; i < n; i++) {
      indexArray[i] = suffix.get(i)[0];
    }
  }

  int search(String key) {
    int left = 0;
    int right = origin.length() - 1;
    while (left <= right) {
      int mid = (left + right) / 2;
      int val = key.compareTo(origin.substring(indexArray[mid], indexArray[mid] + key.length()));
      if (val == 0) return indexArray[mid];
      if (val < 0) right = mid - 1; else left = mid + 1;
    }
    return -1;
  }
}
